pkg/controller/clone-controller.go

```diff
@@ -234,108 +245,376 @@ func (cc *CloneController) waitTargetPodRunningOrSucceeded(pvc *v1.PersistentVol
-		if podSucceededFromPVC(pvc) && pod.Status.Phase == v1.PodRunning {
-			klog.V(3).Infof("Clone succeeded, waiting for source pod %s/%s to stop running", pod.Namespace, pod.Name)
+		if podSucceededFromPVC(pvc) && pod.Status.Phase == corev1.PodRunning {
+			log.V(3).Info("Clone succeeded, waiting for source pod to stop running", "pod.Namespace", pod.Namespace, "pod.Name", pod.Name)
 			return nil
 		}
 
-		if err = cc.clientset.CoreV1().Pods(pod.Namespace).Delete(pod.Name, &metav1.DeleteOptions{}); err != nil {
+		if err = r.Client.Delete(context.TODO(), pod); err != nil {
 			if !k8serrors.IsNotFound(err) {
 				return errors.Wrap(err, "error deleting clone source pod")
 			}
 		}
 	}
 
-	_, err = removeFinalizer(cc.clientset, pvc, cloneSourcePodFinalizer)
+	return r.updatePVC(r.removeFinalizer(pvc, cloneSourcePodFinalizer))
+}
+
+// CreateCloneSourcePod creates our cloning src pod which will be used for out of band cloning to read the contents of the src PVC
+func (r *CloneReconciler) CreateCloneSourcePod(image, pullPolicy, clientName string, pvc *corev1.PersistentVolumeClaim, log logr.Logger) (*corev1.Pod, error) {
+	exists, sourcePvcNamespace, sourcePvcName := ParseCloneRequestAnnotation(pvc)
+	if !exists {
+		return nil, errors.Errorf("bad CloneRequest Annotation")
+	}
+
+	ownerKey, err := cache.MetaNamespaceKeyFunc(pvc)
 	if err != nil {
-		return err
+		return nil, errors.Wrap(err, "error getting cache key")
 	}
 
-	cc.podExpectations.DeleteExpectations(key)
+	clientCert, clientKey, err := r.clientCertGenerator.MakeClientCert(clientName, nil, uploadClientCertDuration)
+	if err != nil {
+		return nil, err
+	}
 
-	return nil
-}
+	serverCABundle, err := r.serverCAFetcher.BundleBytes()
+	if err != nil {
+		return nil, err
+	}
 
-func (cc *CloneController) validateSourceAndTarget(targetPvc *v1.PersistentVolumeClaim) error {
-	sourcePvc, err := getCloneRequestSourcePVC(targetPvc, cc.Controller.pvcLister)
+	podResourceRequirements, err := GetDefaultPodResourceRequirements(r.Client)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	if err = validateCloneToken(cc.tokenValidator, sourcePvc, targetPvc); err != nil {
-		return err
+	pod := MakeCloneSourcePodSpec(image, pullPolicy, sourcePvcName, sourcePvcNamespace, ownerKey, clientKey, clientCert, serverCABundle, pvc, podResourceRequirements)
+
+	if err := r.Client.Create(context.TODO(), pod); err != nil {
+		return nil, errors.Wrap(err, "source pod API create errored")
 	}
 
-	return ValidateCanCloneSourceAndTargetSpec(&sourcePvc.Spec, &targetPvc.Spec)
+	log.V(1).Info("cloning source pod (image) created\n", "pod.Namespace", pod.Namespace, "pod.Name", pod.Name, "image", image)
+
+	return pod, nil
 }
 
-//Run is being called from cdi-controller (cmd)
-func (cc *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
-	cc.Controller.run(threadiness, stopCh, cc.runPVCWorkers)
+func getCloneSourcePodName(pvc *corev1.PersistentVolumeClaim) string {
```

Comment:
You are attempting to create a unique name for the source pod but `findCloneSourcePod` is still looking up by label value.  I initially thought this was not ideal.  But it is good for backward compatibility.

I'm personally not crazy about hashes/UIDs being the the entirety of a resource name.  See the names of replicasets created by deployments for example.  Hash is only part of the name.  But if you want to do something quick and dirty here, I think I would prefer the target PVC UID in place of the hash.  Basically use the `CloneUniqueID` label value.

Reply:
Okay, the goal was a deterministic unique label that would not overflow the max name length. If you prefer the CloneUniqueID over this, that works for me. I will make this function generate that string.